sensors: STM32: Add support for L5 die temp sensor

wcappelle opened the pull request:
This PR adds the different handling of temperature sensor for the
STM32L5 soc. In this soc, there are some calibration settings which
need to be applied for temperature conversion which seem different than other seriers.

Afaik this is the only soc which needs this different handling.

Datasheet snippet for reviewers:
![image](https://user-images.githubusercontent.com/10128093/146527083-3c7bef38-b182-422f-995f-5c081f3cf35f.png)

avisconti reviewed (approved):
LGTM

AdiIntel commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -67,13 +70,24 @@ static int stm32_temp_channel_get(const struct device *dev,
+	temp -= *TEMPSENSOR_CAL1_ADDR;
+	temp *= (TEMPSENSOR_CAL2_TEMP - TEMPSENSOR_CAL1_TEMP);
+	temp /= (*TEMPSENSOR_CAL2_ADDR - *TEMPSENSOR_CAL1_ADDR);
+	temp += 30;
```
Magic number (30). Macro could be better replacement.

wcappelle replied:
done

AdiIntel commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -67,13 +70,24 @@ static int stm32_temp_channel_get(const struct device *dev,
 	}
 
+#ifdef CONFIG_SOC_SERIES_STM32L5X
+	temp = ((float)data->raw * cfg->tsv_mv) / TEMPSENSOR_CAL_VREFANALOG;
```
do you want to assert for a non-zero value of `TEMPSENSOR_CAL_VREFANALOG`

wcappelle replied:
done

erwango commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -11,6 +11,10 @@
 
 LOG_MODULE_REGISTER(stm32_temp, CONFIG_SENSOR_LOG_LEVEL);
 
+#ifdef CONFIG_SOC_SERIES_STM32L5X
+#include <stm32l5xx_ll_adc.h>
+#endif
```
Doesn't seem required. Please remove

wcappelle replied:
done, moved to dts as requested

erwango commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -67,13 +68,25 @@ static int stm32_temp_channel_get(const struct device *dev,
 		return -ENOTSUP;
 	}
 
+#ifdef CONFIG_SOC_SERIES_STM32L5X
+	BUILD_ASSERT(TEMPSENSOR_CAL_VREFANALOG != 0);
+	temp = ((float)data->raw * cfg->tsv_mv) / TEMPSENSOR_CAL_VREFANALOG;
+	temp -= *TEMPSENSOR_CAL1_ADDR;
+	temp *= (TEMPSENSOR_CAL2_TEMP - TEMPSENSOR_CAL1_TEMP);
+	temp /= (*TEMPSENSOR_CAL2_ADDR - *TEMPSENSOR_CAL1_ADDR);
+	temp += TEMPSENSOR_CAL1_TEMP; /* temperature offset: 30°C */
```
I'm puzzled with this section using a different configuration scheme than other series.
I'd prefer to avoid that each new contributor  makes things in his own way

I understand that L5 is a different beast and has to be handled in a different way,
but I had a look and U5 will use the same IP.

So let's let's add a L5 dedicated compatible (that could be reused for U5 later and probably
other series). There are examples in other drivers (STM32 UART would be the most simple)
on how to handle various compatibles in a single driver.

aurel32 replied:
> I'm puzzled with this section using a different configuration scheme than other series. I'd prefer to avoid that each new contributor makes things in his own way

The existing implementation is specific to F1, F4 and F7 families and does not work on other series, like the F0, L4, L5 or WL. This PR is therefore definitely an improvement.

erwango replied:
Maybe I wasn't clear enough.
A dedicated algorithm is required, no discussion about that.
My point is this implementation is done using HAL defines, while initial one is based on device tree.

wcappelle replied:
I will do the changes after my holidays, i agree with you @erwango , just need some look into other drivers what you exactly, i can split the compatible, but I have only knowledge on l5, so can i make it l5 specific compatible? I anyhow need the includes from the hal since this algorithm works with factory calibrated values which i need from the device itself via the hal. Any suggestion would be welcome.

erwango replied:
> so can i make it l5 specific compatible

Sure, others will be able to use the same compatible.

> I anyhow need the includes from the hal since this algorithm works with factory calibrated values which i need from the device itself via the hal. 

Using device tree configuration method, you need to report values from HAL/ref man to device tree.

wcappelle replied:
@erwango, These are not constants for the series, but constants to the device specific. They contain pointer addresses to the calibrated values in production to the single device.  Therefore I cannot use the devicetree for this since you don't know them at compile time. I could add the memory addresses they point to to the devicetree, but I think that would be far from portable and would require too much knowledge of the user if ever needed to be changed

from stm32l5xx_ll_adc.h:
```
/* Temperature sensor */
#define TEMPSENSOR_CAL1_ADDR               ((uint16_t*) (0x0BFA05A8UL)) /* Internal temperature sensor, address of parameter TS_CAL1: On STM32L5, temperature sensor ADC raw data acquired at temperature  30 DegC (tolerance: +-5 DegC), Vref+ = 3.0 V (tolerance: +-10 mV). */
#define TEMPSENSOR_CAL2_ADDR               ((uint16_t*) (0x0BFA05CAUL)) /* Internal temperature sensor, address of parameter TS_CAL2: On STM32L5, temperature sensor ADC raw data acquired at temperature 110 DegC (tolerance: +-5 DegC), Vref+ = 3.0 V (tolerance: +-10 mV). */
#define TEMPSENSOR_CAL1_TEMP               (30L)                        /* Internal temperature sensor, temperature at which temperature sensor has been calibrated in production for data into TEMPSENSOR_CAL1_ADDR (tolerance: +-5 DegC) (unit: DegC). */
#define TEMPSENSOR_CAL2_TEMP               (110L)                       /* Internal temperature sensor, temperature at which temperature sensor has been calibrated in production for data into TEMPSENSOR_CAL2_ADDR (tolerance: +-5 DegC) (unit: DegC). */
#define TEMPSENSOR_CAL_VREFANALOG          (3000UL)                     /* Analog voltage reference (Vref+) voltage with which temperature sensor has been calibrated in production (+-10 mV) (unit: mV). */

```

erwango replied:
What we should store in device tree are not the calibrated values, but the pointer addresses (which are constant for each series, but may vary between series), and all the constants that you just listed.
This doesn't change anything to the algorithm you're using, just the place were you get the constants.

Btw, I had a look to reference manuals for a bunch of series and there are actually several series that could use the variant you're creating. Considering this, it would be more adapted to create a new compatible ("st,stm32-temp-v2" for instance) dedicated to this new variant. This new compatible could include `st,stm32-temp`, block the properties that are not common and add new required properties, cf https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/clock/st%2Cstm32wb-rcc.yaml for an example.
Then, you should use compatibles for configuration, rather than `CONFIG_SOC_SERIES_STM32L5X`.

wcappelle replied:
ok, i've split acually in a seperate compatible with the common parts split off. Let me know if you prefer another way. I think this approach makes it more readible, but if you don't agree, i can change it. Same applies for the compatible name.

erwango commented on dts/bindings/sensor/st,stm32-temp.yaml:
```diff
@@ -28,6 +28,7 @@ properties:
         datasheet "Electrical characteristics/Operating conditions"
         STM32F1 Table 5.3.19 (min 4 mV/C, max 4.6, default 4.3)
         STM32F4 Table 6.3.21 default 2.5
+        Note: This is not used for SOC_SERIES_STM32L5X
```
Let's be consistent and use STM32L5 instead of SOC_SERIES_STM32L5X

erwango replied:
EDIT: Remove this and create a new compatible dedicated to L5 temp sensor variant (and similar)

erwango reviewed (changes requested):
Thanks for this change but I'd like to keep consistency across series.
Also, please add a board overlay in `samples/sensor/stm32_temp_sensor`
so this can be tested in CI (and one can see how to get it working)

erwango commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -67,13 +68,25 @@ static int stm32_temp_channel_get(const struct device *dev,
+	int32_t mv = data->raw * cfg->tsv_mv / 0x0FFF; /* Sensor value in millivolts */
+
 	if (cfg->is_ntc) {
-		temp = (float)(cfg->v25_mv - data->mv);
+		temp = (float)(cfg->v25_mv - mv);
 	} else {
-		temp = (float)(data->mv - cfg->v25_mv);
+		temp = (float)(mv - cfg->v25_mv);
 	}
-	temp = (temp/cfg->avgslope)*10;
+	temp = (temp / cfg->avgslope) * 10;
 	temp += 25;
```
Modifications to existing code should be done in a prelinary commit (so it is easier to review that you're not breaking anything)

wcappelle replied:
done

Commit:
sensors: STM32: Prepare for different handling of temp conversions

moving the conversion from adc value to the get function
which will be used for different handling of stm32 temp sensors

Signed-off-by: Wouter Cappelle <wouter.cappelle@crodeon.com>

Commit:
sensors: STM32: code formatting of stm32_temp with sample

clang format on stm32_temp.c & the sample

Signed-off-by: Wouter Cappelle <wouter.cappelle@crodeon.com>

Commit:
sensors: STM32: Add support for L5 die temp sensor

This PR adds the different handling of temperature sensor for the
STM32L5 soc. In this soc, there are some calibration settings which
need to be applied for temperature conversion.

Signed-off-by: Wouter Cappelle <wouter.cappelle@crodeon.com>

wcappelle commented:
@erwango , all requested changes should have been done, could you do a re-review if you find some time?
Thanks!

erwango commented on dts/bindings/sensor/st,stm32-temp-cal.yaml:
```diff
@@ -0,0 +1,54 @@
+properties:
+    ts-cal1-addr:
+      type: int
+      default: 0x0BFA05A8
```
Please remove default (no reason to privilege L5 over other series), apply everywhere.

wcappelle replied:
done, move them to the l5 board overlay & made the fields required

erwango commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -24,10 +33,17 @@ struct stm32_temp_data {
 	int v25_mv;
-	int tsv_mv;
 	bool is_ntc;
+	bool has_cal;
```
Is that bit strictly required ?
Both configurations are mutually exclusives, so build time condition could be used (`HAS_CALIBRATION`).

wcappelle replied:
done

erwango commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -86,10 +111,17 @@ static const struct sensor_driver_api stm32_temp_driver_api = {
 	struct adc_sequence *asp = &data->adc_seq;
 	int rc;
 
+	if (cfg->has_cal && (!cfg->cal_vrefanalog || !cfg->cal1_addr || !cfg->cal2_addr ||
+			     cfg->cal1_temp >= cfg->cal2_temp)) {
+		LOG_ERR("Invalid configuration");
+		return -EINVAL;
+	}
```
I think that dt tools should prevent misconfiguration, so might be useless.

wcappelle replied:
you still can assign 0 to all ot these, which would case division by 0 or null pointer referenc. The checks are not covered by dt tools, or at leat I couldn't find a range option in the dts bindings, so I think it's safer to keep them if not possible in the dts binding?

erwango reviewed (changes requested):
Thanks for taking into account the comments.
Some changes required, but global direction is ok.

wcappelle commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -64,15 +80,24 @@ static int stm32_temp_channel_get(const struct device *dev, enum sensor_channel
-
-	if (cfg->is_ntc) {
-		temp = (float)(cfg->v25_mv - mv);
+	if (cfg->has_cal) {
```
I was indeed thinking of using more the defines, but that would make the code either work compile time for one or the other compatible. While now it's "ready" to work for more than one compatible at the same time (when reworking DRV_COMPAT in driver config creations ). 
On the other hand you can argue if there would ever be more than 1 temp sensor in the mcu itself... so you have a preference for this?

erwango replied:
Both hardware blocks are mutually exclusive, so let's take it into consideration and use this to get a smaller footprint driver.
(Otherwise current proposal was indeed a good way to handle differences in runtime)

wcappelle replied:
That's clear, thanks for the answer. I've cleaned it up a bit with this info. Also made only 1 instance possible, as it doesn't make sense, and it adds up to the readability

wcappelle commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -24,10 +33,17 @@ struct stm32_temp_data {
 };
 
 struct stm32_temp_config {
```
adds again complexity in pointing & using the correct structure. I think with the current changes, this should be ok? If not, you can reopen

wcappelle commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -116,19 +148,37 @@ static int stm32_temp_init(const struct device *dev)
+	.cal1_temp = DT_INST_PROP(idx, ts_cal1_temp),                                              \
+	.cal2_temp = DT_INST_PROP(idx, ts_cal2_temp),                                              \
+	.cal_vrefanalog = DT_INST_PROP(idx, ts_cal_vrefanalog),                                    \
+	.cal_offset = DT_INST_PROP(idx, ts_cal_offset),
```
done

wcappelle commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -116,19 +148,37 @@ static int stm32_temp_init(const struct device *dev)
+#define STM32_TEMP_INST_COND(idx) STM32_TEMP_INST(idx, STM32_TEMP_CONFIG_NOCAL)
+#endif
+
+DT_INST_FOREACH_STATUS_OKAY(STM32_TEMP_INST_COND)
```
Very good point. I've made only one instance possible. This makes things much simpler to read.

erwango commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -86,10 +113,21 @@ static const struct sensor_driver_api stm32_temp_driver_api = {
 	int rc;
 
+#if HAS_CALIBRATION
+	if (!cfg->cal_vrefanalog || !cfg->cal1_addr || !cfg->cal2_addr ||
```
This first line is superflous.
As these parameters are stated as "required" in bindings, compilation will break if one of these is not given:
```
-- Found devicetree overlay: ~/zephyrproject/zephyr/samples/sensor/stm32_temp_sensor/boards/stm32l562e_dk.overlay
devicetree error: 'ts-cal1-addr' is marked as required in 'properties:' in ~/zephyrproject/zephyr/dts/bindings/sensor/st,stm32-temp-cal.yaml, but does not appear in <Node /stm-temp in '~/zephyrproject/zephyr/misc/empty_file.c'>

```

wcappelle replied:
I tried to answer this one in your previous remark on this line:

> you still can assign 0 to all of these values in the dts, which would cause division by 0 or null pointer reference. The checks are not covered by dt tools, or at least I couldn't find a range option in the dts bindings, so I think it's safer to keep them if not possible in the dts binding?

So this should cover a dts config as
```
/ {
    stm-temp {
        label = "DIE_TEMP";
        compatible = "st,stm32-temp-cal";
        io-channels = <&adc1 17>;
        status = "okay";
        ts-cal1-addr = <0>;
        ts-cal2-addr = <0>;
        ts-cal1-temp = <30>;
        ts-cal2-temp = <110>;
        ts-cal-vrefanalog = <0>;
        ts-cal-offset = <30>;
    };
};
```

I will remove it since it's less likely to make mistakes on these now, since it is moved to the soc dts instead of the user's dts.

erwango commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -67,10 +66,12 @@ static int stm32_temp_channel_get(const struct device *dev,
 		return -ENOTSUP;
 	}
 
+	int32_t mv = data->raw * cfg->tsv_mv / 0x0FFF; /* Sensor value in millivolts */
```
I wonder if this shouldn't be float here ?

wcappelle replied:
I've checked it, and there is no difference with original implementation:

original code:
`data->mv = data->sample_buffer * cfg->tsv_mv / 0x0FFF;`
with `data->sample_buffer` being `int16_t` and `data->mv` being `int32_t`
I think you can loose here 1mv in precision, which is with a 12bit adc with Vref on 3V like 1 LSB.

I rather don't touch this since I cannot test the difference.

erwango commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -116,19 +154,25 @@ static int stm32_temp_init(const struct device *dev)
+#endif
+};
+
+static struct stm32_temp_data stm32_temp_dev_data = { .adc = DEVICE_DT_GET(
+							      DT_INST_IO_CHANNELS_CTLR(0)),
+						      .channel = DT_INST_IO_CHANNELS_INPUT(0) };
```
Please break a line before `.adc...`, since we have plenty of space that's a shame to put all near the 100th column.

wcappelle replied:
I understand what you mean, but this is the output of clang-format with the zephyr .clang-format rules. 
I've added a "," at the end of the .channel assignment, which causes clang-format to do the better formatting.
Hope that's ok for you? Not sure this is allowed by the coding guidelines

erwango replied:
.clang-format rules might be broken for this case.
ok to keep the extra ',', but for next time, don't hesitate to fix things manually.
`.clang-format` rules should be considered as a helper not a rule to apply.

erwango commented on samples/sensor/stm32_temp_sensor/boards/stm32l562e_dk.overlay:
```diff
@@ -0,0 +1,26 @@
+ */
+
+/ {
+    stm-temp {
+        label = "DIE_TEMP";
+        compatible = "st,stm32-temp-cal";
+        io-channels = <&adc1 17>;
+        status = "okay";
+        ts-cal1-addr = <0x0BFA05A8>;
+        ts-cal2-addr = <0x0BFA05CA>;
+        ts-cal1-temp = <30>;
+        ts-cal2-temp = <110>;
+        ts-cal-vrefanalog = <3000>;
+        ts-cal-offset = <30>;
+    };
```
Let's move this node directly in stm32l5.dtsi (with `status = "okay";`) . It will then be accessible to any other application.
In this file, you'll only need to change status node.

wcappelle replied:
good point! 
I think you meant `status = "disabled";` right?

erwango commented on samples/sensor/stm32_temp_sensor/src/main.c:
```diff
@@ -9,11 +9,19 @@
+#elif DT_HAS_COMPAT_STATUS_OKAY(st_stm32_temp_cal)
+#define TEMP_NODE DT_INST(0, st_stm32_temp_cal)
+#else
+#define TEMP_NODE DT_INVALID_NODE
```
Let's use an `#error ` directive here to directly detect configuration issues at build time

wcappelle commented:
@erwango , Thanks for the fast review cycles. I've pushed your last requested changes :-) .

erwango commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -24,10 +33,19 @@ struct stm32_temp_data {
 struct stm32_temp_config {
+	int tsv_mv;
+#if HAS_CALIBRATION == 0
 	int avgslope;
 	int v25_mv;
-	int tsv_mv;
 	bool is_ntc;
+#else
+	uint16_t *cal1_addr;
+	uint16_t *cal2_addr;
+	int cal1_temp;
+	int cal2_temp;
+	int cal_vrefanalog;
+	int cal_offset;
+#endif
```
Nit: instead of negative logic, let's use `#if HAS_CALIBRATION` and keep the same order as you used below.

wcappelle replied:
Is indeed more consistent. Done.

erwango reviewed (changes requested):
Last tiny change requested otherwise we're good.
Thanks for all the effort to make things generic, it seems this will already be useful to others (see https://github.com/zephyrproject-rtos/zephyr/issues/41742).

ycsin commented on drivers/sensor/stm32_temp/stm32_temp.c:
```diff
@@ -67,13 +82,24 @@ static int stm32_temp_channel_get(const struct device *dev,
 	}
-	temp = (temp/cfg->avgslope)*10;
+	temp = (temp / cfg->avgslope) * 10;
 	temp += 25;
```
Is this formula correct? Shouldn't you add 30 instead?
![image](https://user-images.githubusercontent.com/1563974/149173153-d7d6ca84-91b9-4991-9812-93e144c5cc6a.png)
